Implementing stable REST API for Apache Airflow (#274)
apache/airflow-site: landing-pages/site/content/en/blog/implementing-stable-API-for-Apache-Airflow.md

 later discarded this because with our custom date parser we were able to use duration and
 not just date times.

### Other Endpoints
I implemented some different other endpoints. One peculiar issue I faced was because of Marshmallow 2
not giving error when extra fields are in the request body. I implemented a `validate_unknown`
method on the schema to handle this. Thankfully, Flask-AppBuilder updated to using Marshmallow 3,
we quickly updated Flask-AppBuilder in Apache Airflow and started using Marshmallow 3 too.

Here are some PRs I contributed that are related to the REST API:

 1. [Add event log endpoints](https://github.com/apache/airflow/pull/9227)
 2. [Add CRUD endpoints for connection](https://github.com/apache/airflow/pull/9266)
 3. [Add log endpoint](https://github.com/apache/airflow/pull/9331)
 4. [Move limit & offset to kwargs in views plus work on a configurable maximum limit](https://github.com/apache/airflow/pull/9431)
 5. [Update FlaskAppBuilder to v3](https://github.com/apache/airflow/pull/9648)
 6. [Add migration guide from the experimental REST API to the stable REST API](https://github.com/apache/airflow/pull/9771)

> ### Follow-Ups

Reviewer: ```suggestion

### Follow-Ups
```

---
title: "Implementing Stable API for Apache Airflow"
linkTitle: "Implementing Stable API for Apache Airflow"
author: "Ephraim Anierobi"
twitter: "ephraimbuddy"
github: "ephraimbuddy"
description: "An Outreachy intern's progress report on contributing to Apache Airflow REST API."
tags: ["REST API"]
date: "2020-07-19"
---

> My [Outreachy internship](outreachy.org) is coming to its ends which is also the best time to look back and

Author: ```suggestion
My [Outreachy internship](https://outreachy.org) is coming to its ends which is also the best time to look back and
```

reflect on the progress so far.

> The goal of my project is to Extend and Improve the Apache Airflow REST API. In this post,

Reviewer 2: grammar nit - unless "Extend and Improve" are part of a formal title they should be lowercase - "extend and improve" 
If it's the formal title of the project though, that's ok!

Author: Yes. It is the formal title

I will be sharing my progress so far.

We started a bit late implementing the REST API because it took time for the OpenAPI 3.0
specification we were to use for the project to be merged. Thanks to [Kamil](https://github.com/mik-laj),
who paved the way for us to start implementing the REST API endpoints. Below are the endpoints I
implemented and the challenges I encountered, including how I overcame them.

### Implementing The Read-Only Connection Endpoints
The [read-only connection endpoints](https://github.com/apache/airflow/pull/9095) were the first endpoint I implemented. Looking back,
I can see how much I have improved.

> I started by implementing the database schema for the Connection table using Marshmallow 2.

Reviewer 2: add a link to Marshmallow 2 - especially because the word Marshmallow is hard to search (you get lots of food results 😆 - it made me hungry!)

We had to use Marshmallow 2 because Flask-AppBuilder was still using it and Flask-AppBuilder

> is deeply integrated to Apache Airflow. This means I had to unlearn Marshmallow 3 that I have

Reviewer 2: grammar nits - tense agreement -  
* means -> meant 
* have been -> had been

>  been studying before this realization and thankfully, I started using Marshmallow 2 in no time.

Reviewer 2: Consider adjusting the last part of this sentence to explain why you were able to use it in no time - perhaps something like 

"...before this realization, but thankfully, Marshmallow 3 isn't too different, so I was able to start using Marshmallow 2 in no time."

You can swap out "isn't too different" with whatever reason is true for you - "the docs are detailed" "my mentor helped" or whatever it is 😄

This first PR would have been more difficult than it was had it been there was no reference

> endpoint to look at. [Kamil](https://github.com/mik-laj) implemented a draft PR in which I took inspiration from.

Reviewer 2: are you able to link to the draft PR? I think readers would find it interesting to see the inspiration alongside your end product! (which, thank you for linking to your pull above!)

> Thanks to this, It was easy for me to write the unit tests. It was also in this endpoint that

Reviewer 2: Just double checking grammar here - is "parametrized" a particular library? (I genuinely don't know) - if not, did you mean "parameters"? If it's a library, you might want to use the `code` emphasis you used later in the post

>  I learned using parameterized in unit tests :D.

Reviewer 3: ```suggestion
 I learned to use parameterized in unit tests :D.
```

### Implementing The Read-Only DagRuns Endpoints

> This [endpoint](https://github.com/apache/airflow/pull/9153) came with its many challenges, most especially on filtering with `datetimes`.

Reviewer 2: grammar nit - use either "most" or "especially" - together they are redundant

> This was because the connexion library we were using to build the REST API was not validating

Reviewer 2: possible typo - connexion -> connection

Reviewer 2: or this may be a library - in which case, `code` emphasis and a link would help

date-time format in OpenAPI 3.0 specification. This I would later find out, was intentional.

> Connexion dropped `strict-rfc3339` because of the later licence which is not compatible with

Reviewer 2: typo - licence -> license

Reviewer 2: (happens twice)

Apache 2.0 licence.

I implemented a workaround on this, by defining a function called `conn_parse_datetime` in the
API utils module. This was later refactored and thankfully, [Kamil](https://github.com/mik-laj)
 implemented a decorator that allowed us to have cleaner code on the views while using this function.

> We later tried using `rfc3339-validator` whose licence is compatible with Apache 2.0 licence but

Reviewer 2: another licence -> license

(license is one of those horrible words for spelling that I always feel never looks right no matter which way you spell it!! 😄 that and vacuum)

 later discarded this because with our custom date parser we were able to use duration and
 not just date times.

### Other Endpoints

> I implemented some different other endpoints. One peculiar issue I faced was because of Marshmallow 2

Reviewer 2: Yay I love hearing about particular issues and how folks learned from them!!

Reviewer 2: Out of curiosity is this something that was fixed in Marshmallow 3?

Author: Yes. It was fixed in Marshmallow 3

not giving error when extra fields are in the request body. I implemented a `validate_unknown`
method on the schema to handle this. Thankfully, Flask-AppBuilder updated to using Marshmallow 3,
we quickly updated Flask-AppBuilder in Apache Airflow and started using Marshmallow 3 too.

> Here are some PRs I contributed that are related to the REST API:

Reviewer 2: This list is cool! Would you consider adding what the impact these PRs had  on other developers/users? What did it fix/make better/easier/more transparent?

Author: I'm yet to add this. I will add it tomorrow

 1. [Add event log endpoints](https://github.com/apache/airflow/pull/9227)
 2. [Add CRUD endpoints for connection](https://github.com/apache/airflow/pull/9266)
 3. [Add log endpoint](https://github.com/apache/airflow/pull/9331)
 4. [Move limit & offset to kwargs in views plus work on a configurable maximum limit](https://github.com/apache/airflow/pull/9431)
 5. [Update FlaskAppBuilder to v3](https://github.com/apache/airflow/pull/9648)
 6. [Add migration guide from the experimental REST API to the stable REST API](https://github.com/apache/airflow/pull/9771)

### Follow-Ups

> There are still lots of works to be done on the REST API including writing helpful documentation.

Reviewer 2: Grammar nit - change to "There is still lots of work"

---
title: "Implementing Stable API for Apache Airflow"
linkTitle: "Implementing Stable API for Apache Airflow"
author: "Ephraim Anierobi"
twitter: "ephraimbuddy"
github: "ephraimbuddy"
description: "An Outreachy intern's progress report on contributing to Apache Airflow REST API."
tags: ["REST API"]
date: "2020-07-19"
---

My [Outreachy internship](https://outreachy.org) is coming to its ends which is also the best time to look back and
reflect on the progress so far.

The goal of my project is to Extend and Improve the Apache Airflow REST API. In this post,
I will be sharing my progress so far.

We started a bit late implementing the REST API because it took time for the OpenAPI 3.0
specification we were to use for the project to be merged. Thanks to [Kamil](https://github.com/mik-laj),
who paved the way for us to start implementing the REST API endpoints. Below are the endpoints I
implemented and the challenges I encountered, including how I overcame them.

### Implementing The Read-Only Connection Endpoints

> The [read-only connection endpoints](https://github.com/apache/airflow/pull/9095) were the first endpoint I implemented. Looking back,

Reviewer 4: ```suggestion
The [read-only connection endpoints](https://github.com/apache/airflow/pull/9095) was the first endpoint I implemented. Looking back,
```
I'm not sure if it was only one endpoint or more of them that you implemented. If it was only one then the correct grammatical form is "was" :-)

Author: It has two endpoints.

> I can see how much I have improved.

Reviewer 4: It's amazing to hear that! <3

I started by implementing the database schema for the Connection table using [Marshmallow 2](https://marshmallow.readthedocs.io/en/2.x-line/).
We had to use Marshmallow 2 because Flask-AppBuilder was still using it and Flask-AppBuilder
is deeply integrated to Apache Airflow. This meant I had to unlearn Marshmallow 3 that I had
 been studying before this realization, but thankfully, [Marshmallow 3](https://marshmallow.readthedocs.io/en/stable/index.html) isn't too
 different, so I was able to start using Marshmallow 2 in no time.

> This first PR would have been more difficult than it was had it been there was no reference

Reviewer 4: ```suggestion
This first PR would have been more difficult than it was unless there had been any reference
```
I'm not sure if you were choosing between various tenses but atm this sentence seems a bit too complicated and "there" seems unnecessary to me. I was guessing what you were trying to write, so please correct me if I'm wrong with my assumptions.

Author: I am trying to say that there was a reference PR that made it easy for me to write my first PR. That without the reference PR, it would have been hard to make my first PR. I'm not really good at English but working on it :)

Reviewer 4: Don't worry Ephraim, I'm only trying to help here and you are a hard worker from what I've heard, so everything's fine :-) In case of what you've explained, I'd suggest to implement the change I proposed.

Author: Thanks!

endpoint to look at. [Kamil](https://github.com/mik-laj) implemented a [draft PR](https://github.com/apache/airflow/pull/9045) in which I took inspiration from.
Thanks to this, It was easy for me to write the unit tests. It was also in this endpoint that
 I learned using [parameterized](https://github.com/wolever/parameterized) in unit tests :D.

### Implementing The Read-Only DagRuns Endpoints

This [endpoint](https://github.com/apache/airflow/pull/9153) came with its many challenges, especially on filtering with `datetimes`.
This was because the ``connexion`` library we were using to build the REST API was not validating

> date-time format in OpenAPI 3.0 specification. This I would later find out, was intentional.

Reviewer 4: ```suggestion
date-time format in OpenAPI 3.0 specification what I eventually found out, was intentional.
```

Connexion dropped `strict-rfc3339` because of the later license which is not compatible with
Apache 2.0 license.

I implemented a workaround on this, by defining a function called `conn_parse_datetime` in the
API utils module. This was later refactored and thankfully, [Kamil](https://github.com/mik-laj)
 implemented a decorator that allowed us to have cleaner code on the views while using this function.

> We later tried using `rfc3339-validator` whose license is compatible with Apache 2.0 licence but

Reviewer 4: ```suggestion
Then we tried using `rfc3339-validator` whose license is compatible with Apache 2.0 licence but
```
You use a lot of "later" in this paragraph so I added a few words that can replace it. Hope it's fine with you.

 later discarded this because with our custom date parser we were able to use duration and
 not just date times.

### Other Endpoints
I implemented some different other endpoints. One peculiar issue I faced was because of Marshmallow 2
not giving error when extra fields are in the request body. I implemented a `validate_unknown`
method on the schema to handle this. Thankfully, Flask-AppBuilder updated to using Marshmallow 3,
we quickly updated Flask-AppBuilder in Apache Airflow and started using Marshmallow 3 too.

Here are some PRs I contributed that are related to the REST API:

 1. [Add event log endpoints](https://github.com/apache/airflow/pull/9227)
    The event log would help users get information on operations performed at the UI

 2. [Add CRUD endpoints for connection](https://github.com/apache/airflow/pull/9266)
    This PR performs DELETE, PATCH and POST operations on ``Connection``

 3. [Add log endpoint](https://github.com/apache/airflow/pull/9331)

>     This PR enable users to get Task Instances log entries

Reviewer 4: ```suggestion
    This PR enables users to get Task Instances log entries
```

 4. [Move limit & offset to kwargs in views plus work on a configurable maximum limit](https://github.com/apache/airflow/pull/9431)
    This helped us in having a neat code on the views and added configurable maximum limit on query results.

 5. [Update FlaskAppBuilder to v3](https://github.com/apache/airflow/pull/9648)

>     This enabled Airflow to start using v3 of Flask App Builder and also enabled the API to use

Reviewer 4: ```suggestion
    This enabled Airflow to start using v3 of Flask App Builder and also made it possible for the API to use
```
2 x "enabled" 😸

     a modern database serializer/deserializer

 6. [Add migration guide from the experimental REST API to the stable REST API](https://github.com/apache/airflow/pull/9771)

>     This would enable users to start using the stable REST API in less time.

Reviewer 4: Why "would"? Is it because this PR is in progress?

Author: No, but the stable REST API has not been released yet

Reviewer 4: Gotcha, thanks for the explanation :)

### Follow-Ups
There is still lots of works to be done on the REST API including writing helpful documentation.

> I still follow up on these and hopefully, we will complete the REST API before the internship ends.

Reviewer 4: Great news! 🐱

I am very grateful to my mentors, [Jarek](https://github.com/potiuk) and [Kaxil](https://github.com/kaxil) for their
patience with me and for surviving my never-ending questions.

Thank you so much, [Leah E. Cole](https://github.com/leahecole), for your wonderful reviews.

> Thanks for reading...

Reviewer 4: ```suggestion
Thanks for reading! 
```
You should be proud of this blog post because it's great! There's no need to use ellipsis 😸

I can see how much I have improved.

I started by implementing the database schema for the Connection table using [Marshmallow 2](https://marshmallow.readthedocs.io/en/2.x-line/).
We had to use Marshmallow 2 because Flask-AppBuilder was still using it and Flask-AppBuilder
is deeply integrated to Apache Airflow. This meant I had to unlearn Marshmallow 3 that I had
 been studying before this realization, but thankfully, [Marshmallow 3](https://marshmallow.readthedocs.io/en/stable/index.html) isn't too
 different, so I was able to start using Marshmallow 2 in no time.

This first PR would have been more difficult than it was unless there had been any reference
endpoint to look at. [Kamil](https://github.com/mik-laj) implemented a [draft PR](https://github.com/apache/airflow/pull/9045) in which I took inspiration from.
Thanks to this, It was easy for me to write the unit tests. It was also in this endpoint that
 I learned using [parameterized](https://github.com/wolever/parameterized) in unit tests :D.

### Implementing The Read-Only DagRuns Endpoints

This [endpoint](https://github.com/apache/airflow/pull/9153) came with its many challenges, especially on filtering with `datetimes`.

> This was because the ``connexion`` library we were using to build the REST API was not validating

Reviewer: ```suggestion
This was because the `connexion` library we were using to build the REST API was not validating
```
